crimson/heartbeat: report to monitors about osd failure

xxhdx1985126 opened the pull request:
<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
Now the crimson-osd doesn't report osd failures, the pr fix that.
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

tchaikov commented on src/crimson/osd/heartbeat.cc:
```diff
@@ -309,6 +312,29 @@ seastar::future<> Heartbeat::handle_you_died()
+void Heartbeat::heartbeat_check()
+{
+  bool failure = false;
+  for (auto& p: peers) {
```
```suggestion
  for (const auto& [osd, peer_info] : peers) {
```

tchaikov commented on src/crimson/osd/heartbeat.cc:
```diff
@@ -309,6 +312,29 @@ seastar::future<> Heartbeat::handle_you_died()
+    if (clock::is_zero(p.second.first_tx))
+      continue;
+
+    if (p.second.is_unhealthy(clock::now())) {
```
move `clock::now()` out of the loop. and define a constant for it.

tchaikov replied:
@xxhdx1985126 this is not addressed.

tchaikov commented on src/crimson/osd/heartbeat.cc:
```diff
@@ -309,6 +312,29 @@ seastar::future<> Heartbeat::handle_you_died()
+      continue;
+
+    if (p.second.is_unhealthy(clock::now())) {
+      std::time_t deadline = clock::to_time_t(
```
instead of defining a local variable for printing, i'm inclined to reuse the formatter. i extract the one used by messenger in #33849 . could you give it a try?

xxhdx1985126 replied:
Ok, I'll try it:-)

tchaikov replied:
@xxhdx1985126 this is not addressed.

tchaikov commented on src/crimson/osd/heartbeat.cc:
```diff
@@ -309,6 +310,29 @@ seastar::future<> Heartbeat::handle_you_died()
 
+void Heartbeat::heartbeat_check()
+{
+  bool failure = false;
```
```suggestion
  unsigned failures = 0;
```

it'd help with debugging. moreover, i think it's simpler to just remove `failure_queue` member variable and just return the identified osds, and pass them to `send_heartbeats()`.

xxhdx1985126 replied:
I got your point:-) Will try that.

xxhdx1985126 replied:
Um...By send_heartbeats(), what you actually mean is send_failures(), right? Or, are you suggesting merging the send_failures() with send_heartbeats()?

tchaikov replied:
sorry, i meant to say `send_failures()`.

tchaikov commented on src/crimson/osd/heartbeat.cc:
```diff
@@ -309,6 +310,29 @@ seastar::future<> Heartbeat::handle_you_died()
+    }
+  }
+  if (failure)
+    send_failures();
```
```suggestion
  return failures;
```

tchaikov commented on src/crimson/osd/heartbeat.cc:
```diff
@@ -33,7 +33,10 @@ Heartbeat::Heartbeat(const crimson::osd::ShardServices& service,
-    timer{[this] { (void)send_heartbeats(); }}
+    timer{[this] {
+      heartbeat_check();
+      (void)send_heartbeats();
```
```suggestion
      if (auto failures = heartbeat_check(); !failures.empty()) {
        (void)send_heartbeats(std::move(failures);
      }
```

cyx1231st commented on src/crimson/osd/heartbeat.cc:
```diff
@@ -50,9 +50,9 @@ seastar::future<> Heartbeat::start(entity_addrvec_t front_addrs,
+                         SocketPolicy::lossless_peer(0));
   back_msgr->set_policy(entity_name_t::TYPE_OSD,
-                        SocketPolicy::stateless_server(0));
+                        SocketPolicy::lossless_peer(0));
```
I don't think so, this will break the compatibility to be able to talk with classic OSDs.

> this would cause problem
when both sides of a heartbeat connection tries to reconnect, which is not expected for
lossy connections

Could you elaborate what you have observed if it is an issue, does it cause serious bugs?

xxhdx1985126 replied:
stateless_server's connection is lossy, so when both sides tries to reconnect, there's a high possibility that the code goes into handle_existing_connection(), and as the connection is lossy, one of them would be reset and the surviving connection goes into execute_establishing() method in which it tries to register the connection. But if when handling reset, the application tries to reconnect again, then the connection register would trigger the assert "ceph_assert(added);" in SocketMessenger::register_conn.

I agree with on that this breaks compatibility with classic osds, shall I add two hb client msgr just like what the classic osds do? Thanks:-)

cyx1231st replied:
> when handling reset, the application tries to reconnect again, then the connection registering would trigger the assert "ceph_assert(added);" in SocketMessenger::register_conn.

Ahh, it's actually a bug which will also happen elsewhere. The reset event dispatching here should be asynchronous, this connection should be closed/unregistered and replaced before notifying reset and reconnect. Then the reconnect attempt will simply get the replaced connection.

> I agree with you on that this breaks compatibility with classic osds, shall I add two hb client msgrs just like what the classic osds do? Thanks:-)

Let me fix the bug first and see if there are other issues. According to my current understanding, it could be possible to save a connection between every two peering crimson OSDs.

xxhdx1985126 replied:
> EDIT: I think the policy should be SocketPolicy::lossy_client(0), as I see the differences between SocketPolicy::stateless_server(0) are:

Can lossy_client serve as a server?

cyx1231st replied:
> Can lossy_client serve as a server to be connected to?

I think so. The only difference between `SocketPolicy::lossy_client(0)` and `SocketPolicy::stateless_server(0)` is the boolean `Policy::server`, which leads to some limited different behaviors as I listed above according to `src/crimson/net/ProtocolV2.cc`. And it seems to me that `SocketPolicy::stateless_server(0)` is not a correct choice.

Anyway, we can always fall back to the two-connection implementation, if `SocketPolicy::lossy_client(0)` turns out to be impractical.

cyx1231st replied:
Could you split the commit "crimson/heartbeat: avoid heartbeat deadlock on connection reset" into two parts? The policy change is not related to the title.

And if you feel uncertain about the policy change, you can choose not to include it.

xxhdx1985126 replied:
@cyx1231st Ok, I'll split the commit:-)

cyx1231st replied:
Did some tests with `SocketPolicy::lossy_client(0)`, no issue found. See https://github.com/ceph/ceph/pull/33909/commits/096d6d407367211f874f34d336b7c930322a3ca3

tchaikov commented on src/crimson/osd/heartbeat.cc:
```diff
@@ -33,7 +33,10 @@ Heartbeat::Heartbeat(const crimson::osd::ShardServices& service,
     back_msgr{back_msgr},
     // do this in background
-    timer{[this] { (void)send_heartbeats(); }}
+    timer{[this] {
```
nit, might want to revise the title, like

> crimson/osd/heartbeat: report peer failures to monitor

xxhdx1985126 replied:
ok:-)

xxhdx1985126 commented:
There are some bugs in alienstore that need to be fixed before fully testing this PR, please take a look at [33943](https://github.com/ceph/ceph/pull/33943), thanks:-)

cyx1231st commented on src/crimson/osd/heartbeat.cc:
```diff
@@ -199,12 +199,10 @@ seastar::future<> Heartbeat::remove_peer(osd_id_t peer)
+  found->second.con_front->close();
+  found->second.con_back->close();
+  peers.erase(peer);
+  return seastar::now();
```
`Heartbeat::remove_peer()` doesn't need to return `future<>`, And there are chained impacts to other implementations, see https://github.com/ceph/ceph/pull/33909/commits/1d34640a47eb46d7958b5960d8316a89681f8bef.

xxhdx1985126 commented:
@tchaikov @cyx1231st I've modified this pr as suggested, please take a look:-) I've removed the commit that delt with the heartbeat connection deadlock as yingxin has delt with it in [1d34640](https://github.com/ceph/ceph/commit/1d34640a47eb46d7958b5960d8316a89681f8bef)

tchaikov commented on src/crimson/osd/heartbeat.cc:
```diff
@@ -309,6 +312,27 @@ seastar::future<> Heartbeat::handle_you_died()
+					peer_info.last_rx_front);
+    }
+  }
+  if (failure_queue.size())
```
```suggestion
  if (!failure_queue.empty())
```

tchaikov commented on src/crimson/osd/heartbeat.cc:
```diff
@@ -353,7 +377,7 @@ seastar::future<> Heartbeat::send_heartbeats()
 }
 
-seastar::future<> Heartbeat::send_failures()
+seastar::future<> Heartbeat::send_failures(failure_queue_t failure_queue)
```
```suggestion
seastar::future<> Heartbeat::send_failures(failure_queue_t&& failure_queue)
```

xxhdx1985126 commented:
jenkins test make check

tchaikov commented on src/crimson/osd/heartbeat.h:
```diff
@@ -39,7 +45,8 @@ class Heartbeat : public crimson::net::Dispatcher {
 
   seastar::future<> send_heartbeats();
-  seastar::future<> send_failures();
+  seastar::future<> send_failures(failure_queue_t failure_queue);
```
could you move `send_heartbeats()`, `send_failures()` and `heartbeat_check()` down to the `private` section? as they are not part of the public interface of this class. and might want to keep `clock` and `failure_queue_t` close to where they are used.

tchaikov replied:
could you reconsider `heartbeat_check()`?

tchaikov commented on src/crimson/osd/heartbeat.cc:
```diff
@@ -375,7 +399,6 @@ seastar::future<> Heartbeat::send_failures()
                                                   osdmap->get_addrs(osd)});
       return monc.send_message(failure_report);
     }).then([this] {
-      failure_queue.clear();
```
please remove this continuation along with `failure_queue.clear()`, we don't need it anymore.

xxhdx1985126 replied:
Um.. it seems this is not my most recent update..

tchaikov commented on src/crimson/osd/heartbeat.cc:
```diff
@@ -309,6 +312,27 @@ seastar::future<> Heartbeat::handle_you_died()
+    if (peer_info.is_unhealthy(clock::now())) {
+      logger().error(" heartbeat_check: no reply from osd.{} "
+		     "since back {} front {} (oldest deadline {})",
+		     osd, clock::to_time_t(peer_info.last_rx_back),
```
no need to use `clock::to_time_t` for printing `ceph::coarse_real_clock::time_point`. please use `crimson/common/formatter.h` for access its formatter.

xxhdx1985126 commented:
@tchaikov Please take a look:-) Thanks:-)

tchaikov commented on src/crimson/osd/heartbeat.cc:
```diff
@@ -309,6 +312,26 @@ seastar::future<> Heartbeat::handle_you_died()
+					peer_info.last_rx_front);
+    }
+  }
+  if (!failure_queue.empty())
```
please add curly braces.

tchaikov commented:
- https://github.com/ceph/ceph/pull/33836#discussion_r392771614
- https://github.com/ceph/ceph/pull/33836#discussion_r392798063
- https://github.com/ceph/ceph/pull/33836/files#r392798310

please address them.

tchaikov commented on src/crimson/osd/heartbeat.cc:
```diff
@@ -309,6 +312,26 @@ seastar::future<> Heartbeat::handle_you_died()
+    }
+  }
+  if (!failure_queue.empty())
+    send_failures(std::move(failure_queue));
```
if you are ignoring the returned future, please note down the reason. and to be explicit to ignore it.

tchaikov commented on src/crimson/osd/heartbeat.cc:
```diff
@@ -309,6 +312,26 @@ seastar::future<> Heartbeat::handle_you_died()
+		     osd, peer_info.last_rx_back, peer_info.last_rx_front,
+		     peer_info.ping_history.begin()->second.deadline);
+      failure_queue[osd] = std::min(peer_info.last_rx_back,
+					peer_info.last_rx_front);
```
wrong indent.

tchaikov commented on src/crimson/osd/heartbeat.cc:
```diff
@@ -309,6 +312,26 @@ seastar::future<> Heartbeat::handle_you_died()
+  failure_queue_t failure_queue;
+  for (const auto& [osd, peer_info]: peers) {
+    if (clock::is_zero(peer_info.first_tx))
+      continue;
```
could you please add curly braces?

tchaikov commented on src/crimson/osd/heartbeat.cc:
```diff
@@ -309,6 +312,29 @@ seastar::future<> Heartbeat::handle_you_died()
+void Heartbeat::heartbeat_check()
+{
+  failure_queue_t failure_queue;
+  auto now = clock::now()
```
this does not compile.
```suggestion
  const auto now = clock::now();
```

tchaikov commented on src/crimson/osd/heartbeat.cc:
```diff
@@ -309,6 +312,29 @@ seastar::future<> Heartbeat::handle_you_died()
+    }
+  }
+  if (!failure_queue.empty()) {
+    (void)send_failures(std::move(failure_queue));
```
please add comment explaining why we want/can call this method in background.

xxhdx1985126 commented:
@tchaikov please take a look:-) Thanks:-)

tchaikov commented on src/crimson/osd/heartbeat.cc:
```diff
@@ -309,6 +312,34 @@ seastar::future<> Heartbeat::handle_you_died()
+  }
+  if (!failure_queue.empty()) {
+    // send_failures can run in background, because messages
+    // are send in order, if later checks find out the previous
```
s/send/sent/

xxhdx1985126 replied:
corrected:-)

tchaikov commented on src/crimson/osd/heartbeat.cc:
```diff
@@ -309,6 +312,34 @@ seastar::future<> Heartbeat::handle_you_died()
+    // are sent in order, if later checks find out the previous
+    // "failed" peers to be healthy, that "still alive" messages
+    // would be sent after the previous "osd failure" messages
+    // which is totally safe.
```
actually i am more concerned with the life cycle management of the captured `Heartbeat` instance. as once it's captured by a continuation and sent to reactor we have no way to ensure it outlives the capturing continuation, in this case, when `timer` gets triggered, it schedules a continuation to be executed by reactor. but OSD **could** be destroyed before the last chained continuation of the scheduled event is executed by reactor even the timer is removed from reactor by `~timer()`. that's why the tasks scheduled by timer is either atomic or just a single continuation or multiple continuations protected by gate or a semaphore.

xxhdx1985126 replied:
Ah, thanks:-)

tchaikov replied:
@xxhdx1985126 could you please address my comment?

xxhdx1985126 replied:
added comment, please take a look at [34017](https://github.com/ceph/ceph/pull/34017):-)

tchaikov replied:
thank you! will review it shortly.

xxhdx1985126 replied:
You're welcome:-)
